### DISCUSSION TITLE
How to preserve previously selected values in `MultiSelect` which come from API results

### DISCUSSION THREAD

**Discussion by mcmxcdev** [mantinedev/mantine#1744] [Q&A]
I am using the `MultiSelect` component which uses async data from API and then updates the `data` prop.

Unfortunately, I am running into a situation where previously selected values would get lost (when overwriting `data` prop with completely new values from API)

Therefore, I thought that I would be able to solve this by merging old and new data like this (pseudocode):

```js
const { isLoading: isHospitalsLoading } = useQuery(
  ['hospitals', hospitalsSearchTerm],
  async () => {
    const currHospitals = response.map((provider) => {
      return {
        value: `${provider.presentationName} ${provider.id}`,
        label: provider.presentationName,
        location: `${provider.city}, ${provider.state}`,
        term: hospitalsSearchTerm,
      };
    });

    setAvailableHospitals([...prevHospitals, ...currHospitals]);
  }
);


<MultiSelect
  onInput={(event) => {
    setHospitalsSearchTerm(
      (event.target as HTMLInputElement).value
    );
  }}
  data={availableHospitals}
/>
```

The previously selected values are kept like this, but then the problem is that on refetches after search term change, I run into duplicate entries in the dropdown results.

How can I preserve previously selected values from `MultiSelect`, but then show only new dropdown results?

**Comment by rtivital** [comment]
Use controlled component to store selected values in state – https://mantine.dev/core/multi-select/#controlled

**Reply by aldenho52** [reply]
Hey @rtivital I tried this and it didn't work. Despite having controlled state for selected values and search, the selected values are still cleared after typing in new search terms. The state however remains unchanged so it no longer matches what the UI shows.

`export const RecipientsMultiSelect = ({ form }: Props) => {
  const [invitees, setInvitees] = useState<string[]>([]);
  const [searchTerm, setSearchTerm] = useState<string>('');
  const [searchResults, setSearchResults] = useState([]);
  const [loading, setLoading] = useState(false);
  const organization = useRecoilValue(userOrganizations);

  useEffect(() => {
    handleSearchChange();
  }, [searchTerm])

  const handleSearchChange = async () => {
    console.log('handleSearchChange searchTerm', searchTerm);
    if (searchTerm.length === 0 )
      return;

    setLoading(true);
    const usersResponse = await searchUsers(organization.id, searchTerm);
    console.log('usersResponse', usersResponse);
    const results = usersResponse.body.data.map((user: User) => ({
      ...user,
      value: user.email,
      label: `${user.firstName} ${user.lastName}`
    }));
    console.log('results', results);
    setSearchResults(results);
    setLoading(false);
  };

  return (
    <div>
      <MultiSelect
        data={searchResults}
        value={invitees}
        onChange={setInvitees}
        searchable
        clearable
        placeholder="Select invitees"
        rightSection={loading ? <Loader size={16} /> : null}
        onSearchChange={(query) => {
          console.log('query', query);
          setSearchTerm(query);
          setInvitees(invitees);
        }}
        searchValue={searchTerm}
        nothingFound="Nothing found"
        maxDropdownHeight={250}
      />
      <Autocomplete data={searchResults} multiple={true} />
      <Select data={searchResults} multiple={true} />
    </div>

  );
};`
